perf_benchmark/README.md

```diff
@@ -42,9 +42,145 @@ Options meaning:
+defined protobuf message (`NestedPayload` in `benchmark.proto`) or
+using `google/protobuf/struct.proto`. We benchmarked the effects of having
+nested layers during the transcoding.
+
+- The performance of `google/protobuf/struct.proto` is worse than a recursively
+  defined protobuf message in both JSON -> gRPC and gRPC -> JSON cases.
+- Transcoding for gRPC -> JSON has a much better performance.
+- Transcoding streamed messages doesn't add extra overhead.
+- The per-byte latency of nested structure does not conform to a trend.
+
+![Number of Nested Layers Visualization](image/nested_layers.jpg "Number of Nested Layers")
+
+### Array Length
+
+We benchmarked the effects of varying array length during transcoding.
+
+- Transcoding for JSON -> gRPC has much worse performance than transcoding for
+  gRPC -> JSON.
+- The per-byte latency for non-streaming message converges when the array length
+  is over 1024 - 0.03 us for JSON -> gRPC and 0.001 for gRPC -> JSON.
+- Streaming messages has almost little overhead.
+
+![Array Length Visualization](image/array_length.jpg "Array Length")
+
+### Body Length
+
+We benchmarked the effects of a message containing a single `bytes` typed field
+with varying data length.
+
+- Transcoding for JSON -> gRPC has worse performance than transcoding for gRPC
+  -> JSON.
+- The per-byte latency for non-streaming message converges to 0.01 us for JSON
+  -> gRPC and 0.005 us for gRPC -> JSON.
+- Streaming messages has almost little overhead.
+
+_Note: The JSON representation of `bytes` typed protobuf field is encoded in
+base64. Therefore, 1 MiB sized message in gRPC would have roughly 1.33 MiB in
+JSON. The per-byte latency is calculated using the unencoded data size, which is
+why the per-byte latency would give around 34000 for 32 MiB of data, whereas the
+message latency for 32 MiB is actually around 50000._
+
+![Body Length Visualization](image/body_length.jpg "Body Length")
+
+### Number of Message Segments
+
+We benchmarked the effects for messages arrive in segments. This only applies to
+JSON -> gRPC since gRPC doesn't support incomplete messages. Currently, the
+caller needs to make sure the message arrives in full before transcoding from
+gRPC.
+
+- There is a noticeable increase when the number of message segment increase to
+  more than 1.
+- The overhead scales up linearly as the number of message segments increases.
+- The effect of having segmented message becomes less when the message is
+  streamed.
+
+![Number of Message Segments Visualization](image/num_message_segment.jpg "Number of Message Segments Visualization")
+
+### Value Data Type
+
+We benchmarked transcoding from an array of zeros `[0, 0, ..., 0]` into `string`
+, `double`, and `int32` type.
+
+- `string` data type has the less overhead.
+- `double` has the most significant overhead for transcoding.
+
+![Value Data Type Visualization](image/value_data_type.png "Value Data Type")
```

Comment:
Any idea why double for gRPC -> JSON is an outlier? I would think the data type shouldn't make a big difference

Reply:
Yeah, I wouldn't think it will make any difference either. Let me look into it

Reply:
This is caused by different wiring formats. Please see my latest README for this.

Reply:
Cool. Is supporting packed formats something we can do in the transcoder? Maybe file a P3 bug internally to track this?

Reply:
SG. Created a ticket to track this b/249583469